server-master/src/main/java/com/aliyun/emr/rss/service/deploy/master/MasterUtil.java

```diff
@@ -17,133 +17,386 @@
+      totalAllocations += Math.pow(1 + diskGroupGradient, diskGroups - 1 - i);
+    }
+    for (int i = 0; i < diskGroups; i++) {
+      taskAllocationRatio[i] =
+          Math.pow(1 + diskGroupGradient, diskGroups - 1 - i) / totalAllocations;
+    }
+    logger.info("load-aware offer slots algorithm init with taskAllocationRatio {}",
+        StringUtils.join(taskAllocationRatio, ','));
+    initialized = true;
+  }
+
+  private static List<List<DiskInfo>> placeDisksToGroups(List<DiskInfo> usableDisks,
+      int diskGroupCount) {
+    List<List<DiskInfo>> diskGroups = new ArrayList<>();
+    usableDisks.sort((o1, o2) ->
+                         Math.toIntExact(o1.avgFlushTime() - o2.avgFlushTime()));
+    int diskCount = usableDisks.size();
+    int startIndex = 0;
+    int groupSizeSize = (int) Math.ceil(usableDisks.size() / (double) diskGroupCount);
+    for (int i = 0; i < diskGroupCount; i++) {
+      List<DiskInfo> diskList = new ArrayList<>();
+      if (startIndex >= usableDisks.size()) {
+        continue;
       }
-      masterLocation = new PartitionLocation(
-          partitionIds.get(idx),
-          newEpoch,
-          workers.get(nextMasterInd).host(),
-          workers.get(nextMasterInd).rpcPort(),
-          workers.get(nextMasterInd).pushPort(),
-          workers.get(nextMasterInd).fetchPort(),
-          workers.get(nextMasterInd).replicatePort(),
-          PartitionLocation.Mode.Master,
-          slaveLocation
-      );
-      if (shouldReplicate) {
-        slaveLocation.setPeer(masterLocation);
+      if (startIndex + groupSizeSize <= diskCount) {
+        diskList.addAll(usableDisks.subList(startIndex, startIndex + groupSizeSize));
+        startIndex += groupSizeSize;
+      } else {
+        diskList.addAll(usableDisks.subList(startIndex, usableDisks.size()));
+        startIndex = usableDisks.size();
       }
+      diskGroups.add(diskList);
+    }
+    return diskGroups;
+  }
 
-      // add master location to WorkerInfo
-      locations._1.add(masterLocation);
+  private static Map<WorkerInfo, List<DiskUsableInfo>> getRestriction(List<List<DiskInfo>> groups,
+      Map<DiskInfo, WorkerInfo> diskWorkerMap, int required) {
+    int groupSize = groups.size();
+    long[] groupAllocations = new long[groupSize];
+    Map<WorkerInfo, List<DiskUsableInfo>> restrictions = new HashMap<>();
+    long[] diskGroupTotalSlots = new long[groupSize];
+    for (int i = 0; i < groupSize; i++) {
+      for (DiskInfo disk : groups.get(i)) {
+        diskGroupTotalSlots[i] += disk.availableSlots();
+      }
+    }
+    double[] currentAllocation = new double[groupSize];
+    double currentAllocationSum = 0;
+    for (int i = 0; i < groupSize; i++) {
+      if (!groups.get(i).isEmpty()) {
+        currentAllocationSum += taskAllocationRatio[i];
+      }
+    }
+    for (int i = 0; i < groupSize; i++) {
+      if (!groups.get(i).isEmpty()) {
+        currentAllocation[i] = taskAllocationRatio[i] / currentAllocationSum;
```

Comment:
I suppose it's OK to use taskAllocationRatio directly instead of currentAllocation, since sum of taskAllocationRatio equals 1.